### ISSUE TITLE
Port OW to HAIKU OS

### ISSUE THREAD

**Issue by jmalak** [open-watcom/open-watcom-v2#138]

**Comment by jmalak** [comment]
Hi,
I am trying to get Open Watcom running on the Haiku operating system (www.haiku-os.org). I have fixed some initial issues, and currently I'm stuck with this problem:

    /sources/openwatcom-2.0.0/bld/orl/omf/c/omfmunge.c:89:5: warning: implicit declaration of function 'stricmp' [-Wimplicit-function-declaration]

Haiku has strcasecmp. I see there is a condition on _BSD_SOURCE to use that at the top of this file. However I don't know where I should be adding the _BSD_SOURCE define. Is there a way to add it to the host cflags somehow in the build system?
(I am using gcc and have set OWTOOLS=GCC in setvars.sh)

**Comment by jmalak** [comment]
Thanks for bug reporting.
It is now fixed in git repository 7d23ba88ff2a81660b2a48e4b73d2fcbcb36b837

**Comment by jmalak** [comment]
Hello again,
This seems to work and the build gets further.

Here are my changes so far: https://bitbucket.org/haikuports/haikuports/src/8dd8acc035c080fe894b74923eb1af4c08640d58/sys-devel/openwatcom/patches/openwatcom-2.0.0.patchset
(you probably don't want to apply all of it, yet, but some parts may be relevant).

My _cmdname implementation returns the full path to the executable and seems to work ok. However, bwasm fails to run because it can't open its messages (in the MsgInit function). I have tried to uncomment the -DUSE_TEXT_MSGS flag in master.mif for wasm, which seems to work. However build then fails in a similar way when running bwlink.

On Haiku exectuables are ELF files. While we do support something similar to Windows resources (https://haiku-os.org/legacy-docs/bebook/BResources.html), I don't think Watcom is making use of that. How are resources handled in that case and what is the right way to build the bootstrap tools (bwasm, bwlink) with debug information, so I can investigate this more easily?

**Comment by jmalak** [comment]
If I understand Haiku correctly then Haiku is not UNIX derivate.
It is hard for me to say how big effort will be necessary to port Open Watcom to Haiku, because my knowledge about Haiku is zero.
OW is ported to Linux, but not to BSD (it can be only run as Linux executable on BSD over Linux compatible layer).
OW can be build on POSIX systems but there can be some specific problems, which need fixing.
For OW is Haiku new OS and it requires to create full build environment for new OS (local.mif make file).
What filesystem separators Haiku use? Is filesystem case sensitive or in-sensitive?
What C/C++ compiler is used as system compiler?
For bootstrap tools build clibext.c/clibext.h must be ported to Haiku OS.

Anyway debug version of bootstrap tools can be build by command
builder boot OWDEBUGBUILD=1
in appropriate project directory, by example for WASM it must be run in bld/wasm.

**Comment by jmalak** [comment]
Hi,
Haiku is not (directly) UNIX derivative but it is POSIX compatible. I am one of the main developers of Haiku, but have little knowledge about the OpenWatcom build system. We should be able to work together on this :). My porting attempt so far is defining __UNIX__ (which gets most of the clibext building) and overriding parts that create problems as I find them (for example the /proc filesystem doesn't exist on Haiku so I had to rewrite _cmdname as you can see in the patch).

The filesystem uses / as a separator and is case sensitive. I'm using gcc 4.8.4 (we also have gcc 2.95.3 but I think it is better to not try using that).

Should I create a separate issue for this, or is it fine to continue here?

**Comment by jmalak** [comment]
First task is port WMAKE to HAIKU OS.
Please, what gcc macros identify HAIKU OS?

**Comment by jmalak** [comment]
Next question, what is output of "uname" command for HAIKU systems.

**Comment by pulkomandy** [comment]
We define __HAIKU__.

I had a look at why bwasm was failing to load its resources, the problem is that off_t in Haiku is 64-bit (even on 32-bit systems). The code sign-extends some negative seek values by casting them to long, which is only 32-bits (for example in bld/wres/c/loadfind.c FindResources). This give a big positive number instead of the expected one and seek fails.

**Comment by jmalak** [comment]
I need identify HAIKU OS automaticaly by gcc macro as on Linux macro __linux__ is defined etc.
For shell script build.sh I need to know how HAIKU is identified by uname command.

Anyway, thanks for analysis of WASM problem.

**Comment by pulkomandy** [comment]
As I said the gcc macro to use is `__HAIKU__` (it is defined by default by the gcc preprocessor)
uname -o and -s output "Haiku"

**Comment by jmalak** [comment]
Thanks for your info.
I submit all necesary changes to OW build system.
Now you should create your own copy of setvars.sh where you do your local change for OWTOOL etc.
Next try to run clean.sh shell script to cleanup build tree
next run build.sh script to build OW standard way

**Comment by pulkomandy** [comment]
I tried your changes but there is still missing things. I made some extra fixes and now I'm back to my problem with bwasm failing to read its resources.

Current version of the patch:
http://bb.haikuports.org/haikuports/src/871cecde3edf8c6fe0d820053ba8337f995fa9f5/sys-devel/openwatcom/patches/openwatcom-2.0.0.patchset?at=master

Relevant parts:
* build/mif/local.mif copypaste error, Haiku definitions still using BSD variables in some places
* bld/builder/master.mif, bld/wl/wlobjs.mif: use POSIX sources for Haiku

**Comment by jmalak** [comment]
Thanks for patch.
I have a question why you can not use local macro and replace it by static?

**Comment by jmalak** [comment]
Problem with seeking function prototype mismatch is now fixed in git repository by 5a32e0ec5f87c05e52bbb61592fce46995b1def2 this caused problem on HAIKU OS with access to text resources by wres library

**Comment by jmalak** [comment]
Your patch is applied  8aff3cbcbcf8f78ef316b4887c0c552ca028a821

**Comment by pulkomandy** [comment]
Hi,
I'm not sure about the static/local change, maybe this is not needed anymore (I had to do that a long time ago when trying to port openwatcom 1.9). I will try to build without this.

I tried a build with all the new changes. I had to fix some more things (definitions of FileOffset in wl, as and bwasm still using the long type, mainly). But bwasm still fails to load  the resources. I will debug it again to try to see what's wrong now.

**Comment by jmalak** [comment]
To local problem. In OW1.9 it was problem but in V2 for it should be OK to use local where it is used now.
I don't study any code using seeking in application, but only fix resource manipulation library (wres) which you report as problem. Unfortunately now can appear next problems masked by this bug.
Note me if you have any trouble. Now we have basic configuration of building system for HAIKU OS and corrected make files for building bootstrap tools on HAIKU OS. Now should be only problems with source code portability.

**Comment by pulkomandy** [comment]
Ok, I made some progress.
After fixing the build issues I got much further, and then hit another problem. When building bld/win386/conv/win386, there is a step involving running conv.exe to build win38.lib. But conv.exe is a PE executable for Windows, not an ELF file, so it can't be run.
Build issues I needed to fix is some local/static confusion (I tried to remove my previous patch but I got errors, so I did similar changes. Maybe only the function declarations need to be "fixed"?), and the seeking problems already mentionned, but in the tools: http://bb.haikuports.org/haikuports/src/4c2614bf6599740ef7b4cba8953148cc1562efac/sys-devel/openwatcom/patches/openwatcom-2.0.0.patchset?at=master

**Comment by jmalak** [comment]
OK it looks like local macro is used or defined  somewhere in HAIKU or GCC headers.
I will completely remove local macro.
I will fix win386 library build, it looks like some macro in make file is not defined for HAIKU OS.

**Comment by jmalak** [comment]
You have incorrect changes in local.mif which is for BSD, probably you use __BSD__ macro, but it is wrong. __BSD__ macro is reserved macro used for BSD OS  by build system (by wmake or by compilers). You can define only one OS macro, exception is __UNIX__ macro which is generaly used to define UNIX like filesystem (case sensitive and '/' as path separator, also define list separator etc.) as oposite for DOS-ish filesystems (case insensitive and '\' as path separator, also define list separator etc.).
Problem with win386 library build is strange. Please, could you send me appropriate build log with env.variable OWVERBOSE=1 for win386 library build.

**Comment by pulkomandy** [comment]
The changes in local.mif are needed. I'm not using `__BSD__` here. The problem is when you introduced haiku support , it sets `common_cppflags_haiku` but then sets other variables using `common_cppflags_bsd` which isn't set. The result is that `__UNIX__` does not get set and build fails. My changes are in the haiku section of local.mif, I didn't touch the BSD part.


Building win386 ( the failing part):
```
BUILD-chroot:/sources/openwatcom-2.0.0/bld/win386/conv/win386> ../../../../build/bin/builder build OWVERBOSE=1               
**** BUILD rule
================ 18:15:25 /sources/openwatcom-2.0.0/bld/win386 ================
========== 18:15:25 /sources/openwatcom-2.0.0/bld/win386/conv/win386 ==========
bwcl386 -zq    -I"/sources/openwatcom-2.0.0/bld/hdr/dos"  -I"/sources/openwatcom-2.0.0/bld/watcom/h" ../c/conv.c -ox -w8 -fo=.obj -x -xx -zam -fe=./conv.exe
Warning! W1093: cannot open /sources/openwatcom-2.0.0/bld/wstuba/dosi86/wstub.exe
bwasm -q -we -d1 -DNDEBUG -fo=fldenv.obj -3-w8 -bt=windows       -I"/sources/openwatcom-2.0.0/bld/hdr/dos" -I"/sources/openwatcom-2.0.0/bld/w16api/wini86" -I"/sources/openwatcom-2.0.0/bld/watcom/h" ../asm/fldenv.asm
bwasm -q -we -d1 -DNDEBUG -fo=_call16.obj -3-w8 -bt=windows       -I"/sources/openwatcom-2.0.0/bld/hdr/dos" -I"/sources/openwatcom-2.0.0/bld/w16api/wini86" -I"/sources/openwatcom-2.0.0/bld/watcom/h" ../asm/_call16.asm
./conv.exe ../def/win386 ../def -l -s -q
sh: ./conv.exe: cannot execute binary file: Not an executable
Error(E42): Last command making (win386.lib) returned a bad status
Error(E02): Make execution terminated
<pmake -d build         -h> => non-zero return: 2
Build failed
```

conv.exe is a windows executable in PE format (it starts with MZ instead of the ELF signature), which I can't run. It seems bwcl386 is not built with correct settings or wrong command line is passed.

**Comment by pulkomandy** [comment]
builder in bld/wcl/386:

```
BUILD-chroot:/sources/openwatcom-2.0.0/bld/wcl/386/binbuild> ../../../../build/bin/builder build OWVERBOSE=1
**** BUILD rule
=============== 18:24:21 /sources/openwatcom-2.0.0/bld/wcl/386 ================
============ 18:24:21 /sources/openwatcom-2.0.0/bld/wcl/386/dosi86 ============
bwcc -zq -w8 -we -oaxt -d1 -x -xx -zam -fo=clcommon.obj -j -fpi -ml -bt=dos  -s    -DNDEBUG  -D__WATCOM_LFN__ -DWCL386    -I"/sources/openwatcom-2.0.0/bld/trmem" -s -I. -I.. -I"../../h"  -I"/sources/openwatcom-2.0.0/bld/hdr/dos"  -I"/sources/openwatcom-2.0.0/bld/watcom/h" ../../c/clcommon.c
bwcc -zq -w8 -we -oaxt -d1 -x -xx -zam -fo=cmdlhelp.obj -j -fpi -ml -bt=dos  -s    -DNDEBUG  -D__WATCOM_LFN__ -DWCL386     -I. -I.. -I"../../h"  -I"/sources/openwatcom-2.0.0/bld/hdr/dos"  -I"/sources/openwatcom-2.0.0/bld/watcom/h" /sources/openwatcom-2.0.0/bld/watcom/c/cmdlhelp.c
bwcc -zq -w8 -we -oaxt -d1 -x -xx -zam -fo=wcl.obj -j -fpi -ml -bt=dos  -s    -DNDEBUG  -D__WATCOM_LFN__ -DWCL386     -I. -I.. -I"../../h"  -I"/sources/openwatcom-2.0.0/bld/hdr/dos"  -I"/sources/openwatcom-2.0.0/bld/watcom/h" ../../c/wcl.c
bwlink op q name wcl386.exe debug dwarf op nored op symfile op map sys dos lib doslfnl.lib   file {clcommon.obj cmdlhelp.obj wcl.obj }
=========== 18:24:21 /sources/openwatcom-2.0.0/bld/wcl/386/linux386 ===========
bwcc386 -zq -w8 -we -oaxt -d1 -x -xx -zam -fo=clcommon.obj -j -fpi -mf -zc -bt=linux  -s    -DNDEBUG   -DWCL386    -I"/sources/openwatcom-2.0.0/bld/trmem" -s -I. -I.. -I"../../h"  -I"/sources/openwatcom-2.0.0/bld/hdr/linux"  -I"/sources/openwatcom-2.0.0/bld/watcom/h" ../../c/clcommon.c
bwcc386 -zq -w8 -we -oaxt -d1 -x -xx -zam -fo=cmdlhelp.obj -j -fpi -mf -zc -bt=linux  -s    -DNDEBUG   -DWCL386     -I. -I.. -I"../../h"  -I"/sources/openwatcom-2.0.0/bld/hdr/linux"  -I"/sources/openwatcom-2.0.0/bld/watcom/h" /sources/openwatcom-2.0.0/bld/watcom/c/cmdlhelp.c
bwcc386 -zq -w8 -we -oaxt -d1 -x -xx -zam -fo=wcl.obj -j -fpi -mf -zc -bt=linux  -s    -DNDEBUG   -DWCL386     -I. -I.. -I"../../h"  -I"/sources/openwatcom-2.0.0/bld/hdr/linux"  -I"/sources/openwatcom-2.0.0/bld/watcom/h" ../../c/wcl.c
bwlink op q name wcl386.exe debug dwarf op nored op symfile op map sys linux   file {clcommon.obj cmdlhelp.obj wcl.obj }
=========== 18:24:21 /sources/openwatcom-2.0.0/bld/wcl/386/linuxx64 ===========
============ 18:24:21 /sources/openwatcom-2.0.0/bld/wcl/386/nt386 =============
bwcc386 -zq -w8 -we -oaxt -d1 -x -xx -zam -fo=clcommon.obj -j -fpi -mf -zc -bt=nt -DWINVER=0x400 -D_WIN32_IE=0x300  -s    -DNDEBUG   -DWCL386    -I"/sources/openwatcom-2.0.0/bld/trmem" -s -I. -I.. -I"../../h"  -I"/sources/openwatcom-2.0.0/bld/hdr/dos" -I"/sources/openwatcom-2.0.0/bld/w32api/nt" -I"/sources/openwatcom-2.0.0/bld/watcom/h" ../../c/clcommon.c
bwcc386 -zq -w8 -we -oaxt -d1 -x -xx -zam -fo=cmdlhelp.obj -j -fpi -mf -zc -bt=nt -DWINVER=0x400 -D_WIN32_IE=0x300  -s    -DNDEBUG   -DWCL386     -I. -I.. -I"../../h"  -I"/sources/openwatcom-2.0.0/bld/hdr/dos" -I"/sources/openwatcom-2.0.0/bld/w32api/nt" -I"/sources/openwatcom-2.0.0/bld/watcom/h" /sources/openwatcom-2.0.0/bld/watcom/c/cmdlhelp.c
bwcc386 -zq -w8 -we -oaxt -d1 -x -xx -zam -fo=wcl.obj -j -fpi -mf -zc -bt=nt -DWINVER=0x400 -D_WIN32_IE=0x300  -s    -DNDEBUG   -DWCL386     -I. -I.. -I"../../h"  -I"/sources/openwatcom-2.0.0/bld/hdr/dos" -I"/sources/openwatcom-2.0.0/bld/w32api/nt" -I"/sources/openwatcom-2.0.0/bld/watcom/h" ../../c/wcl.c
genverrc ../../exever.rc exever.grc $ wcl386 0
bwrc -q -xb -r -bt=nt -DWINVER=0x400 -D_WIN32_IE=0x300 -D__NT__ -D_WIN32 -D_M_IX86    -I"/sources/openwatcom-2.0.0/bld/hdr/dos" -I"/sources/openwatcom-2.0.0/bld/w32api/nt" -I"/sources/openwatcom-2.0.0/bld/watcom/h" exever.grc -fo=exever.res
bwlink op q name wcl386.exe debug dwarf op nored op symfile op map sys nt   file {clcommon.obj cmdlhelp.obj wcl.obj }
bwrc -q -k exever.res wcl386.exe
============ 18:24:21 /sources/openwatcom-2.0.0/bld/wcl/386/ntx64 =============
============ 18:24:21 /sources/openwatcom-2.0.0/bld/wcl/386/os2386 ============
bwcc386 -zq -w8 -we -oaxt -d1 -x -xx -zam -fo=clcommon.obj -j -fpi -mf -zc -bt=os2  -s    -DNDEBUG   -DWCL386    -I"/sources/openwatcom-2.0.0/bld/trmem" -s -I. -I.. -I"../../h"  -I"/sources/openwatcom-2.0.0/bld/hdr/dos" -I"/sources/openwatcom-2.0.0/bld/os2api/os2386/h" -I"/sources/openwatcom-2.0.0/bld/watcom/h" ../../c/clcommon.c
bwcc386 -zq -w8 -we -oaxt -d1 -x -xx -zam -fo=cmdlhelp.obj -j -fpi -mf -zc -bt=os2  -s    -DNDEBUG   -DWCL386     -I. -I.. -I"../../h"  -I"/sources/openwatcom-2.0.0/bld/hdr/dos" -I"/sources/openwatcom-2.0.0/bld/os2api/os2386/h" -I"/sources/openwatcom-2.0.0/bld/watcom/h" /sources/openwatcom-2.0.0/bld/watcom/c/cmdlhelp.c
bwcc386 -zq -w8 -we -oaxt -d1 -x -xx -zam -fo=wcl.obj -j -fpi -mf -zc -bt=os2  -s    -DNDEBUG   -DWCL386     -I. -I.. -I"../../h"  -I"/sources/openwatcom-2.0.0/bld/hdr/dos" -I"/sources/openwatcom-2.0.0/bld/os2api/os2386/h" -I"/sources/openwatcom-2.0.0/bld/watcom/h" ../../c/wcl.c
bwlink op q name wcl386.exe debug dwarf op nored op symfile op map sys os2v2   file {clcommon.obj cmdlhelp.obj wcl.obj }
=============== 18:24:22 /sources/openwatcom-2.0.0/bld/wcl/386 ================
BUILD-chroot:/sources/openwatcom-2.0.0/bld/wcl/386/binbuild> ls
bwcl386.exe  bwcl386.map  clcommon.obj  cmdlhelp.obj  wcl.obj
```

Are all the versions (I guess for different target architectures) generated in the same folder (binbuild)? If so, I would end up with only the os2 one as they all have the same name. That would explain why I get PE executables out of it, I would need the linux version instead (so I get ELF)?

**Comment by jmalak** [comment]
OK, I understand.
The problem is that first you must have working bootstrap build properly.
You run builder for standard build of OW, but it is too early to run this phase, because other configuration must be done before this phase.
Anyway best will be to work with builder tool directly not by shell scripts.
Now please run clean.sh script to clean all builded objects and get build system to well defined state.
Please check build/bin directory, there should not be any executable or data files.
Next start build.sh script again, wait short time after it start build OW tools and break it.
go to bld directory and run 
builder bootclean
and next
builder boot
This must go smoothly and create full set of build tools in build/bin. probably it will fail to build conv utility in win386 project.
you can repeate after bug fixing
builder bootclean
and
builder boot
until all will be build correctly don't run builder build, it has no sense for this phase.

**Comment by jmalak** [comment]
I corrected make files for regular OW build on HAIKU OS by d3c2d2d188b13c8b1d3f2c5c7df61a273202824b
that now regular build (second phase) should go OK

**Comment by pulkomandy** [comment]
On Tue, Jan 20, 2015 at 09:52:46AM -0800, Jiří Malák wrote:
> OK, I understand.
> The problem is that first you must have working bootstrap build properly.
> You run builder for standard build of OW, but it is too early to run this phase, because other configuration must be done before this phase.
> Anyway best will be to work with builder tool directly not by shell scripts.
> Now please run clean.sh script to clean all builded objects and get build system to well defined state.
> Please check build/bin directory, there should not be any executable or data files.

ok.

> Next start build.sh script again, wait short time after it start build OW tools and break it.

I did this manually, and compiled wmake, then builder.

> go to bld directory and run 
> builder bootclean

ok

> and next
> builder boot
> This must go smoothly and create full set of build tools in build/bin. probably it will fail to build conv utility in win386 project.

This works without problems until the end:

```
Copying '/sources/openwatcom-2.0.0/bld/dmpobj/binbuild/dmpobj.exe' to
'/sources/openwatcom-2.0.0/build/bin/bdmpobj'...
================ 19:51:04 /sources/openwatcom-2.0.0/bld/dmpobj ================
=================== 19:51:04 /sources/openwatcom-2.0.0/bld ====================
```

If running it again it does notthing. So I think I'm past the bootstrap
step now. If I read build.sh correcly the next step is to use 
"builder build", which I did, and only then I got it to fail.

-- 
Adrien.

**Comment by jmalak** [comment]
OK. Did you tried builder build after my last fix?
If any error then could you put here appropriate part of log file.
For testing is better  to set env.variable OWVERBOSE=1 to get full command line in log file.

**Comment by pulkomandy** [comment]
Your fix is working, now I get a crash in bwlink while building bld/nwlib/nt386.dll. I will investigate this.
